File: csharp/1353.maximum-number-of-events-that-can-be-attended.cs
/*
 * @lc app=leetcode id=1353 lang=csharp
 *
 * [1353] Maximum Number of Events That Can Be Attended
 */

using System;
using System.Collections.Generic;
using System.Linq;

// @lc code=start
public class Solution {
    public int MaxEvents(int[][] events) {
        int n = events.Length, ret = 0, i = 0, d = 1, md = events.Max(x => x[1]);
        var sortedEvents = events.Select((x, i) => (End: x[1], Start: x[0], Id: i)).OrderBy(x => x.Start).ToArray();
        var possible = new SortedSet<(int End, int Start, int Id)>(Comparer<(int End, int Start, int Id)>.Create((x, y) => {
            if (x.End != y.End) {
                return x.End.CompareTo(y.End);
            }
            if (x.Start != y.Start) {
                return x.Start.CompareTo(y.Start);
            }
            return x.Id.CompareTo(y.Id);
        }));
        while (d <= md) {
            if (i < n && possible.Count == 0) {
                d = sortedEvents[i].Start;
            }
            while (possible.Count > 0 && possible.Min.End < d) {
                possible.Remove(possible.Min);
            }
            while (i < n && sortedEvents[i].Start == d) {
                possible.Add(sortedEvents[i++]);
            }
            if (possible.Count > 0) {
                possible.Remove(possible.Min);
                ++ret;
            } else if (i >= n) {
                break;
            }
            ++d;
        }
        return ret;
    }
}
// @lc code=end

File: java/1353.maximum-number-of-events-that-can-be-attended.java
/*
 * @lc app=leetcode id=1353 lang=java
 *
 * [1353] Maximum Number of Events That Can Be Attended
 */

import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.Queue;

// @lc code=start

class Solution {
    public int maxEvents(int[][] events) {
        int n = events.length, i = 0, ret = 0, d = 1;
        Queue<Integer> ends = new PriorityQueue<>();
        Arrays.sort(events, java.util.Comparator.comparing((int[] arr) -> arr[0]));
        while (i < n || !ends.isEmpty()) {
            while (!ends.isEmpty() && ends.peek() < d) {
                ends.poll();
            }
            while (i < n && events[i][0] == d) {
                ends.offer(events[i][1]);
                ++i;
            }
            if (!ends.isEmpty()) {
                ends.poll();
                ++ret;
            }
            ++d;
        }
        return ret;
    }
}
// @lc code=end

